### PR TITLE
[RISCV] Support CoreV SIMD builtins in Clang 

### PR DESCRIPTION
**Pull request by melonedo**
This PR adds on top of #27 support for clang to use `__builtin_riscv_cv_simd_xxx` functions. To avoid merge conflicts, a copy of #27 is also included here temporarily before it is merged.

In this PR, I defined Clang builtins according to [corev-builtin-spec.md](https://github.com/openhwgroup/core-v-sw/blob/9bbbf289957542cb4ccd7e8139e5cefec6c6262d/specifications/corev-builtin-spec.md). Their signature should match the specification. To support these builtin functions in a straightforward way in LLVM, I also defined synonymous LLVM intrinsics with format `llvm.riscv.cv.simd.xxx`. However, LLVM does not seem very happy with i16 or i8 arguments which do not perfectly fit in a machine register, so all i8 or i16arguments are zero-extended to i32 before the corresponding LLVM intrinsic is called. 

With the constraint of llvm *lit*, the tests for Clang only reach LLVM IR level, which is not ideal. Although it can be manually checked they do generate CoreV instructions here, I believe tests that compile to assembly are needed as https://github.com/openhwgroup/corev-llvm-project/issues/28#issuecomment-1428203145 points out.

### CODE REVIEW

**Comment by ChunyuLiao** [comment]
If possible, add a test to this file. https://github.com/llvm/llvm-project/blob/main/clang/test/Driver/riscv-arch.c

**Comment by jeremybennett** [comment]
> If possible, add a test to this file. https://github.com/llvm/llvm-project/blob/main/clang/test/Driver/riscv-arch.c

This file is specifically to test that the march arguments are properly recognized (so in this case `xcvsimd`).

File: llvm/include/llvm/IR/IntrinsicsRISCV.td [hunk 1]
```diff
@@ -1582,3 +1582,118 @@ def int_riscv_sm4ed      : ScalarCryptoByteSelectAny;
...
+    def int_riscv_cv_simd_ # NAME # _b : ScalarCoreVSimdGprGprIntrinsic;
+  } else {
+    defm NAME : ScalarCoreVSimdGprGprIntrinsicHB;
+  }
+  defm NAME # _sc : ScalarCoreVSimdGprGprIntrinsicHB;
+}
+
+let TargetPrefix = "riscv" in {
+defm add : CoreVSimdBinary<true>;
+def int_riscv_cv_simd_add_h : ScalarCoreVSimdGprGprImmIntrinsic;
```
Comment (lewis-revill) [inline llvm/include/llvm/IR/IntrinsicsRISCV.td]: For the intrinsics which match Clang builtins, you can add `ClangBuiltin<"__<intrinsic>">` to the definitions of those intrinsics, and I believe this allows us to remove the code added in `CGBuiltin.cpp` as Clang will then automatically match the builtins to the LLVM intrinsics.

For example `def int_riscv_cv_simd_add_h : ScalarCoreVSimdGprGprImmIntrinsic, ClangBuiltin<"__builtin_riscv_cv_simd_add_h">;`

Whether this is worth the effort or not I'm not sure, especially given that we have the multiclasses to think about.

File: llvm/lib/Support/RISCVISAInfo.cpp [hunk 2]
```diff
@@ -103,6 +103,8 @@ static const RISCVSupportedExtension SupportedExtensions[] = {
     {"zicbop", RISCVExtensionVersion{1, 0}},
 
     {"svnapot", RISCVExtensionVersion{1, 0}},
+
```
Comment (lewis-revill) [inline llvm/lib/Support/RISCVISAInfo.cpp]: This should be a separate patch with a separate test, which this patch will then depend on, so we can focus on getting the extension to be parsed correctly on the command line in a different discussion.

File: llvm/lib/Target/RISCV/RISCVInstrInfoCOREV.td [hunk 3]
```diff
@@ -69,6 +69,19 @@ def cv_uimm12 : Operand<XLenVT>,
   let OperandNamespace = "RISCVOp";
 }
 
+def cv_tsimm6 : Operand<XLenVT>, TImmLeaf<XLenVT, [{return isInt<6>(Imm);}]> {
```
Comment (lewis-revill) [inline llvm/lib/Target/RISCV/RISCVInstrInfoCOREV.td]: What's the difference between this and `simm6` and why is it needed?

**Comment by lewis-revill** [comment]
Tests look good and this is a good chunk of builtins added to start catching up with GCC. Please take a look at my review comments though.

**Comment by ChunyuLiao** [comment]
I know how it works( --target=riscv32-unknown-linux-gnu -march=rv32gc_xcvsimd0p1 ) I know how it works. This pr, I look good now

**Comment by simonpcook** [comment]
Hi, after the conversation earlier I wanted to clear up any confusion about the version number things for anyone who was there, in particular when it came to testing; in case we were talking about two different things.

I haven't built LLVM with this patch, but it looks like all the version number stuff is correct, but for Clang it's useful to know that unlike GCC, `-march` is only a driver option, and when calling the underlying compiler `clang -cc1` or assembler `clang -cc1as` the driver will break the `-march` string down into the underlying LLVM `TargetFeatures` which by convention do not have version numbers. As it looks like the tests here do, we call the compiler (`%clang_cc1` is the variable `lit` substitutes), and provide the target feature `+xcvsimd`.

Looking at the patch it looks like you're already following the right convention, but hopefully this is useful context for anyone not familiar with LLVM looking to review it from the CORE-V perspective.

**Comment by melonedo** [comment]
Aligning with newest spec and GCC turned out to be a major change...
Now the builtin is in sync, except:
- Negative bit indices are treated as syntactic error in LLVM but not GCC (e.g. in [cv-simd-extractu-h-compile-1.c](https://github.com/openhwgroup/corev-gcc/blob/02e102db40cef5fa14e59b538d4a681b20511f8c/gcc/testsuite/gcc.target/riscv/cv-simd-extractu-h-compile-1.c)), see https://github.com/openhwgroup/corev-gcc/issues/26.
- I haven't figured out how to implement `cv.shuffleiX.sci.b`, which splits the input constant, any ideas? (Implemented now)

**Comment by ChunyuLiao** [comment]
cv.shuffleiX.sci.b 
https://github.com/openhwgroup/core-v-sw/blob/master/specifications/corev-builtin-spec.md#uint32_t-__builtin_riscv_cv_simd_shuffle_sci_b-uint32_t-i-const-uint8_t-flgs
I think we should use PseudoInsts, than expand them at RISCVExpandPseudoInsts.cpp. 
To build a new imm, I think llvm::APInt maybe required .https://llvm.org/doxygen/classllvm_1_1APInt.html

File: llvm/lib/Target/RISCV/RISCVInstrInfoCOREV.td (hunk 3 above)
Comment (melonedo) [reply llvm/lib/Target/RISCV/RISCVInstrInfoCOREV.td]: The difference is that `cv_tsimm6` is a `TIMMLeaf` while `simm6` is an `IMMLeaf`, which corresponds to `TargetConstantSDNode` instead of `ConstantSDNode`

File: clang/include/clang/Basic/BuiltinsRISCVCOREV.def [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+// This file defines the CORE-V-specific builtin function database.  Users of
+// this file must define the BUILTIN macro to make use of this information.
+//
+//===----------------------------------------------------------------------===//
+
+#if defined(BUILTIN) && !defined(TARGET_BUILTIN)
+#   define TARGET_BUILTIN(ID, TYPE, ATTRS, FEATURE) BUILTIN(ID, TYPE, ATTRS)
+#endif
+
+TARGET_BUILTIN(simd_add_h, "UZiUZiUZiIUc", "nc", "xcvsimd")
```
Comment (simonpcook) [inline clang/include/clang/Basic/BuiltinsRISCVCOREV.def:18]: Do all of these builtins actually need `int32_t` as the types, even on `rv64`?

File: llvm/include/llvm/IR/IntrinsicsRISCV.td (hunk 1 above)
Comment (melonedo) [reply llvm/include/llvm/IR/IntrinsicsRISCV.td]: Currently generating the builtin definitions will not be ideal since it still needs to perform a conversion to `i32` in `CGBuiltin.cpp`.

File: llvm/lib/Support/RISCVISAInfo.cpp (hunk 2 above)
Comment (melonedo) [reply llvm/lib/Support/RISCVISAInfo.cpp]: Fixed in https://github.com/openhwgroup/corev-llvm-project/pull/36

**Comment by melonedo** [comment]
Since this PR is too complicated, I split part of it into #36 and #37. Hope this will make it easier to interpret.

File: llvm/lib/Support/RISCVISAInfo.cpp (hunk 2 above)
Comment (ChunyuLiao) [reply llvm/lib/Support/RISCVISAInfo.cpp]: If this is a separate patch, perhaps clang's builtin support could also be a separate patch, so that no dependencies are needed and each patch is independent.

File: clang/include/clang/Basic/BuiltinsRISCVCOREV.def (hunk 4 above)
Comment (melonedo) [reply clang/include/clang/Basic/BuiltinsRISCVCOREV.def:18]: Do you mean the scope of these builtins should be restricted to `xcvsimd,32bit`?

**Comment by lewis-revill** [comment]
@jeremybennett this PR can be closed since it was split into two separate PRs

File: llvm/lib/Support/RISCVISAInfo.cpp (hunk 2 above)
Comment (jeremybennett) [reply llvm/lib/Support/RISCVISAInfo.cpp]: Already merged

File: clang/include/clang/Basic/BuiltinsRISCVCOREV.def (hunk 4 above)
Comment (jeremybennett) [reply clang/include/clang/Basic/BuiltinsRISCVCOREV.def:18]: This is resolved. It is a matter for when the 64-bit builtins are defined.

**Comment by jeremybennett** [comment]
This has been tested, and found correct on local machines.

File: llvm/include/llvm/IR/IntrinsicsRISCV.td (hunk 1 above)
Comment (jeremybennett) [reply llvm/include/llvm/IR/IntrinsicsRISCV.td]: We have agreed that the current approach is the best in this case.

**Comment by jeremybennett** [comment]
Good to merge.
